src/monitors/BytecodeProfilingMonitor.v3

```diff
@@ -0,0 +1,450 @@
+					if (stack.empty()) {
+						b.put1("0,0,00,%d,", cost).ln();
+					} else {
+						b.put2("%d,0,00,%d,", stack.peek(), cost).ln();
+					}
+				}
+			}
+			rev = rev.tail;
+		}
+		System.write(fd, b.toString());
+	}
+
+	def reportRecoveredBytecodes() {
+		var fd = System.fileOpen(Strings.format1("%s_recovered_bytecodes.csv", output_prefix), false);
+		var function_ids = Maps.keyList(recovered_bytecodes);
+		var b = StringBuilder.new();
+		b.puts("function_id,bytecode\n"); // header
+		while (function_ids != null) {
+			var function_id = function_ids.head;
+			b.put1("%d,", function_id);
+			var bytecodes = recovered_bytecodes[function_id];
+			var bytecode_addrs = Maps.keyList(bytecodes);
+			var min_addr = u32.max;
+			var max_addr: u32 = 0;
+			while (bytecode_addrs != null) {
+				var addr = bytecode_addrs.head;
+				if (addr < min_addr) min_addr = addr;
+				if (addr > max_addr) max_addr = addr;
+				bytecode_addrs = bytecode_addrs.tail;
+			}
+
+			for (i = min_addr; i <= max_addr; i++) {
+				b.put1("/%x", bytecodes[i]);
+			}
+			b.ln();
+			function_ids = function_ids.tail;
+		}
+		System.write(fd, b.toString());
+	}
+
+	private def commonSetup() {
+		def cost = get_cost();
+		match (last_probe_event) {
+			OPCODE_READ(bytecode, offset) => {
+				trace = 
+					List.new(TraceEntry.STEP(bytecode, offset, cost, cur_wasm_func_trace), trace);
+				hotness.record(bytecode, cost);
+			}
+			_ => trace = List.new(TraceEntry.SYSTEM(cost, cur_wasm_func_trace), trace);
+		}
+		last_probe_event_count = u32.!(count_probe.count);
+	}
+
+	private def get_cost() -> u32 {
+		return u32.!(count_probe.count) - last_probe_event_count;
+	}
+}
+
+private class BytecodeHotness(output_prefix: string) {
+	var bytecode_hotness: Array<HashMap<u32, u32>> = Array.new(256);
+
+	def record(b: u8, cost: u32) {
+		if (bytecode_hotness[b] == null) {
+			bytecode_hotness[b] = HashMap<u32, u32>.new(int.!<u32>, u32.==);
+		}
+		if (!bytecode_hotness[b].has(cost)) {
+			bytecode_hotness[b][cost] = 0;
+		}
+		bytecode_hotness[b][cost]++;
+	}
+
+	def update(b: u8, old_cost: u32, new_cost: u32) {
+		if (bytecode_hotness[b] == null || !bytecode_hotness[b].has(old_cost)) {
+			return;
+		}
+		bytecode_hotness[b][old_cost]--;
+		bytecode_hotness[b][new_cost] = bytecode_hotness[b][new_cost] + 1;
+	}
+
+	def report() {
+		var fd = System.fileOpen(Strings.format1("%s_global_hotness.csv", output_prefix), false);
+		var b = StringBuilder.new();
+		b.puts("bytecode,cost,count\n"); // header
+		for (i = 0; i < bytecode_hotness.length; i++) {
+			if (bytecode_hotness[i] != null) {
+				var costs = Maps.keyList(bytecode_hotness[i]);
+				while (costs != null) {
+					var j = costs.head;
+					if (bytecode_hotness[i][j] > 0)
+						b.put3("%x,%d,%d", i, j, bytecode_hotness[i][j]).ln();
+					costs = costs.tail;
+				}
+			}
+		}
+		System.write(fd, b.toString());
+	}
+}
+
+private class CombinedCountProbe(all_counters: List<DynamicLoc -> Resumption>) extends CountProbe {
```

Comment:
I wasn't quite sure why we need a combined counter; AFAICT this is just a count of all the `br_table` executions?